multicluster/controllers/multicluster/serviceexport_controller.go

```diff
@@ -45,18 +142,470 @@ type ServiceExportReconciler struct {
+		}
+		re.Labels["sourceKind"] = common.ServiceKind
+	case common.EndpointsKind:
+		re.ObjectMeta.Name = resName
+		re.Spec.Endpoints = &mcsv1alpha1.EndpointsExport{
+			Subsets: common.FilterEndpointSubsets(ep.Subsets),
+		}
+		re.Labels["sourceKind"] = common.EndpointsKind
+	}
+	return *re
+}
+
+func (r *ServiceExportReconciler) updateOrCreateResourceExport(resName string,
+	ctx context.Context,
+	req ctrl.Request,
+	newResExport *mcsv1alpha1.ResourceExport,
+	existResExport *mcsv1alpha1.ResourceExport,
+	rc core.RemoteCommonArea) error {
+	createResExport := reflect.DeepEqual(*existResExport, mcsv1alpha1.ResourceExport{})
+	resNamespaced := types.NamespacedName{Namespace: leaderNamespace, Name: resName}
+	if createResExport {
+		newResExport.Finalizers = []string{common.ResourceExportFinalizer}
+		klog.InfoS("creating ResourceExport", "resourceexport", resNamespaced.String())
+		err := rc.Create(ctx, newResExport, &client.CreateOptions{})
+		if err != nil {
+			klog.ErrorS(err, "fail to create ResourceExport in leader cluster", "resourceexport", resNamespaced.String())
+			return err
+		}
+	} else {
+		newResExport.ObjectMeta.ResourceVersion = existResExport.ObjectMeta.ResourceVersion
+		newResExport.Finalizers = existResExport.Finalizers
+		err := rc.Update(ctx, newResExport, &client.UpdateOptions{})
+		if err != nil {
+			klog.ErrorS(err, "fail to update ResourceExport", "resourceexport", resNamespaced.String())
+			return err
+		}
+	}
+
+	var message string
+	if createResExport {
+		message = "creation is successful"
+	} else {
+		message = "update is successful"
+	}
+
+	// ignore status update failure?
+	latestResExport := &mcsv1alpha1.ResourceExport{}
+	err := rc.Get(ctx, resNamespaced, latestResExport)
+	if err != nil {
+		klog.InfoS("fail to get ResourceExport", "resourceexport", resNamespaced.String(), "err", err)
+		return nil
+	}
+	latestResExport.Status.Conditions = []mcsv1alpha1.ResourceExportCondition{{
+		Type:               mcsv1alpha1.ResourceExportSucceeded,
+		Status:             corev1.ConditionTrue,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+		Message:            message,
+	}}
+	err = rc.Status().Update(ctx, latestResExport, &client.UpdateOptions{})
+	if err != nil {
+		klog.InfoS("fail to update ResourceExport Status", "resourceexport", resNamespaced.String(), "err", err)
+	}
+	return nil
+}
+
+func getEndPointsAddress(ep *corev1.Endpoints) []string {
+	var epAddrs []string
+	for _, s := range ep.Subsets {
+		for _, a := range s.Addresses {
+			epAddrs = append(epAddrs, a.IP)
+		}
+	}
+	return epAddrs
+}
+
+func getEndPointsPorts(ep *corev1.Endpoints) []corev1.EndpointPort {
+	var epPorts []corev1.EndpointPort
+	for _, s := range ep.Subsets {
+		epPorts = append(epPorts, s.Ports...)
+	}
+	return epPorts
+}
+
+func getResourceExportName(clusterID string, req ctrl.Request, kind string) string {
+	return clusterID + "-" + req.Namespace + "-" + req.Name + "-" + kind
+}
+
+func getPointer(str string) *string {
```

Comment:
Name it getStringPointer(). I start to think maybe you should define global variables for such strings.